src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,218 @@
+
+    private static readonly string[] IceChunksConfigurations =
+        ["iceShard", "iceChunkSmall", "iceChunkBig", "iceSnowflake"];
+
+    [JsonProperty]
+    private readonly XoShiRo256starstar random;
+
+    [JsonProperty]
+    private readonly bool hasEventAlreadyHappened;
+
+    [JsonProperty]
+    private readonly Dictionary<int, Dictionary<Compound, float>> previousEnvironmentalChanges = new();
+
+    [JsonProperty]
+    private readonly Dictionary<int, string> previousBackground = new();
+
+    [JsonProperty]
+    private readonly Dictionary<int, Color> previousLightColour = new();
+
+    /// <summary>
+    /// Tells how many generations the event will last. "-1" means that it hasn't started at all.
+    /// "0" means it has finished, and it won't happen again
+    /// </summary>
+    [JsonProperty]
+    private int generationsLeft = -1;
+
+    [JsonProperty]
+    private GameWorld targetWorld;
+
+    public GlobalGlaciationEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public GlobalGlaciationEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld() { }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        if (generationsLeft > 0)
+            generationsLeft -= 1;
+
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (hasEventAlreadyHappened)
+            return;
+
+        if (generationsLeft == -1)
+            TryToTriggerEvent(totalTimePassed);
+        else if (generationsLeft == 0)
+            FinishEvent();
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (Constants.SurfaceBiomes.Contains(patch.BiomeType))
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlaciationEvent, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        generationsLeft = random.Next(2, 6);
+
+        foreach (var (index, patch) in targetWorld.Map.Patches)
+        {
+            if (Constants.SurfaceBiomes.Contains(patch.BiomeType))
+            {
+                ChangePatchProperties(index, patch, totalTimePassed);
+            }
+        }
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (!Constants.SurfaceBiomes.Contains(patch.BiomeType))
+                continue;
+
+            var oxygenLevel = patch.Biome.ChangeableCompounds[Compound.Oxygen];
```

Comment:
If this doesn't want to change the compound amounts, this should use the proper interface for reading the compounds (`BiomeConditions.TryGetCompound`).